# SPRING 2021
# Applied Machine Learning, Test 2

**Problem 1** 
1. Run the code cell below to load data for this problem (do not make any changes to the content of the cell!). The code creates a new data frame `bc` with some breast cancer data.

In [751]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

data_bc=load_breast_cancer()
bc=pd.DataFrame(data=data_bc['data'], 
                columns=data_bc['feature_names']
                )
bc=bc.iloc[:, :5]
bc['target']=data_bc['target']

2. What is the shape of the created data frame `bc`?

In [752]:
bc.shape

(569, 6)

3. Print to the screen the first 10 rows of the data frame.

In [753]:
bc[:10]

mean radius  mean texture  ...  mean smoothness  target
0        17.99         10.38  ...          0.11840       0
1        20.57         17.77  ...          0.08474       0
2        19.69         21.25  ...          0.10960       0
3        11.42         20.38  ...          0.14250       0
4        20.29         14.34  ...          0.10030       0
5        12.45         15.70  ...          0.12780       0
6        18.25         19.98  ...          0.09463       0
7        13.71         20.83  ...          0.11890       0
8        13.00         21.82  ...          0.12730       0
9        12.46         24.04  ...          0.11860       0

[10 rows x 6 columns]

4. Print to the screen the column names of `bc`.

In [754]:
bc.columns

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'target'],
      dtype='object')

5. Demonstrate that the only values that appear in the `target` column are 0 (malignant) and 1 (benign).

In [755]:
bc['target'].unique()

array([0, 1])

6. Demonstrate that there are no missing values in the `bc` data frame.

In [756]:
bc.isna().sum()

mean radius        0
mean texture       0
mean perimeter     0
mean area          0
mean smoothness    0
target             0
dtype: int64

7. Using one of the standard Pandas functions, create a summary showing basic aggregated statistics (such as the counts, means, standard deviations, minimum and maximum values, 25%, 50%, and 75% percentiles) for all numerical columns of the data frame.

In [757]:
bc.describe()

mean radius  mean texture  ...  mean smoothness      target
count   569.000000    569.000000  ...       569.000000  569.000000
mean     14.127292     19.289649  ...         0.096360    0.627417
std       3.524049      4.301036  ...         0.014064    0.483918
min       6.981000      9.710000  ...         0.052630    0.000000
25%      11.700000     16.170000  ...         0.086370    0.000000
50%      13.370000     18.840000  ...         0.095870    1.000000
75%      15.780000     21.800000  ...         0.105300    1.000000
max      28.110000     39.280000  ...         0.163400    1.000000

[8 rows x 6 columns]

8. Using one of the standard Pandas function, create a summary for the `target` column showing the total numbers of instances for each unique value of `target` (the counts of unique values). 

In [758]:
bc['target'].value_counts()

1    357
0    212
Name: target, dtype: int64

9. Create a boolean mask for the instances whose `'mean area'` values are between 800 and 840. Assume that both limits are inclusive.

In [759]:
import numpy as np

bc_mask = np.bitwise_and(bc['mean area'] >= 800, bc['mean area'] <= 840)

10. Using the boolean mask you created in the previous step, count the number of instances with the `'mean area'` values between 800 and 840.

In [760]:
bc[bc_mask]['mean area'].count()

13

11. Using the same mask and `bc`, count the number of positive class (benign) instances with the `'mean area'` values between 800 and 840.

In [761]:
(bc[bc_mask]['target'] == 1).sum()

3

12. For the instances with the `'mean area'` values between 800 and 840, print to the screen the content of the following columns of `bc` (in this order): `'mean perimeter'`, `'mean radius'`, `'mean area'`, `'target'`.

In [762]:
print(bc[bc_mask]['mean perimeter'],bc[bc_mask]['mean radius'], bc[bc_mask]['mean area'], bc[bc_mask]['target'])

34     107.0
75     104.1
132    106.2
141    105.1
177    109.3
283    108.8
328    106.9
329    107.5
363    106.6
406    104.3
417    102.9
479    109.8
508    104.7
Name: mean perimeter, dtype: float64 34     16.13
75     16.07
132    16.16
141    16.11
177    16.46
283    16.24
328    16.27
329    16.26
363    16.50
406    16.14
417    15.50
479    16.25
508    16.30
Name: mean radius, dtype: float64 34     807.2
75     817.7
132    809.8
141    813.0
177    832.9
283    805.1
328    813.7
329    826.8
363    838.1
406    800.0
417    803.1
479    815.8
508    819.8
Name: mean area, dtype: float64 34     0
75     0
132    0
141    0
177    0
283    0
328    0
329    0
363    1
406    1
417    0
479    0
508    1
Name: target, dtype: int64


13. Create a 50-50 split of the data into training and testing sets using a standard function from the Scikit-Learn library. Set `random_state=42` and perform startification on the target variable. The output of this operation must be a set of four NumPy arrays: `X_train`, `X_test`, `y_train`, `y_test`.

In [763]:
from sklearn.model_selection import train_test_split
X = bc.loc[:'target'].values
print(X.shape)
y = (bc['target']).values
print(y.shape)
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.5, train_size=.5, random_state=42, stratify = y)

(569, 6)
(569,)


14. Demonstrate that the percentages of positive class instances are very close for the training and testing set targets (don't expect a perfect agreement but the percentages should be reasonably close).

In [764]:
print(y_test.size)
print(y_test.shape)
print(y_train.size)
print(X_train.shape)
print(X_test.shape)

285
(285,)
284
(284, 6)
(285, 6)


15. Create a pipeline that scales the data first using the standard scaler method and then performs logistic regression.

In [765]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

pipe = Pipeline([
                 ('sc',StandardScaler()),
                 ('lr', LogisticRegression())
])


16. Using the training data set, compute a 5-fold cross-validated accuracy score for this pipeline. Print to the screen the average accuracy across all 5 folds.

In [766]:
from sklearn.model_selection import cross_val_score

score = cross_val_score(pipe, X_train, y_train, cv=5, scoring='accuracy')
print(f'average accuracy along the five folds are {score.mean()}')

average accuracy along the five folds are 1.0


17. Now, fit the pipeline that you created in step 15 on the training set data and then make predictions on the test set data. After that compute and print to the screen the accuracy of your predictions on the test set data.

In [767]:
from sklearn.metrics import accuracy_score
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test)
acc = accuracy_score(y_train, y_pred)
print(f"the accuracy is {acc:.5f}")

ValueError: ignored

18. Print to the screen the confusion matrix computed on the test set data for the classifier built in the preivous step.

In [ ]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_true=y_train, y_pred=y_pred))

19. What are the number of true positives (TP), true negatives (TN), false positives (FP), and false negatives (FN)?

  **Answer here:**

  * TP = 106
  * TN = 178
  * FP = 0
  * FN = 0

20. Perform all necessary steps to get values of precision and recall for different values of decision threshold. Then plot both the precision and recall values versus the threshold values on the same set of axes. Create meaningful labels, title, choose a reasonable font size, and create a legend indicating which curve is which (precision or recall).

In [ ]:
from sklearn.metrics import precision_score, recall_score
import matplotlib.pyplot as plt



**Problem 2** 
1. Run the code cell below to load data for this problem (do not make any changes to the content of the cell!). The code creates a new data frame `wine` with some wine data. The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine.

In [ ]:
from sklearn.datasets import load_wine

data_wine=load_wine()
wine=pd.DataFrame(data=data_wine['data'], 
                  columns=data_wine['feature_names']
                  )
wine['target']=data_wine['target']
wine.head()

2. Print to the screen the shape of the data frame.

In [ ]:
wine.shape

3. Print to the screen the column names of the `wine` data frame.

In [ ]:
wine.columns

3. Demonstate that the target column contains three classes: 0, 1, and 2 (one class per cultivator); also, count the number of instances of each class (the count of unique values). *Hint:* you can do all this with a single Pandas function.

In [ ]:
wine['target'].value_counts()

4. Demonstrate that there are no missing values in the `wine` data frame.

In [ ]:
wine.isna().sum()

5. Preform a 70-30 split of the data set into a training set and a test set, stratifying on the target, and setting `random_state=42`.

In [ ]:
X = wine.loc[:'target'].values
y = (wine['target']==1).values

X_train,X_test,y_train,y_test = train_test_split(X,y, random_state=42, stratify=y)

6. Build a two-step pipeline that applies the standard scaler to the features and then performs K-Nearest Neighbors classification (do not specify `n_neighbors`).

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
pipe_two = Pipeline([
                     ('sc',StandardScaler()),
                     ('kn', KNeighborsClassifier())
])

7. Use the training set to perform a 5-fold cross-validated grid search for the best number of nearest neighbors $K$ to be used in the pipeline created in the previous step. Explore the values of $K$ from 1 to 70 and set `scoring='f1_macro'` (so that you evaluate the performance based on the macro averaged F1 score).

In [ ]:
from sklearn.model_selection import GridSearchCV

GridSearchCV(pipe_two, param_grid = {'n_neighbors' in range(1,70)},scoring='f1_macro',)

8. Now use the fitted grid serach object from the previous step to make prediction on the test set. Compute the macro averaged F1 score for this predictions and print the result to the screen. *Hint:* set `average='macro'` when computing the macro averaged F1 score.

9. Build another two-step pipeline that applies the standard scaler to the features and then performs a Softmax regression.

In [ ]:
pipe_sf = Pipeline([
                    ('sc', StandardScaler()),
                    ('smax', LogisticRegression(multi_class='multinomial'))
])

10. Fit the Softmax regression pipeline on the training data and make predictions on the test data; then compute the macro averaged F1 score and print it to the screen.

In [769]:
from sklearn.metrics import f1_score

pipe_sf.fit(X,y)

y_pred = pipe_sf.predict(X_test)

f1_score(y_true=y_test, y_pred=y_pred)

1.0